ambari-server/src/main/java/org/apache/ambari/server/agent/stomp/HostLevelParamsHolder.java

```diff
@@ -119,23 +121,42 @@ protected HostLevelParamsUpdateEvent getEmptyData() {
 
   @Subscribe
   public void onClusterComponentsRepoUpdate(ClusterComponentsRepoChangedEvent clusterComponentsRepoChangedEvent) throws AmbariException {
-    updateDataOfCluster(clusterComponentsRepoChangedEvent.getClusterId());
+    Cluster cluster = clusters.getCluster(clusterComponentsRepoChangedEvent.getClusterId());
+    for (Host host : cluster.getHosts()) {
+      updateDataOfHost(clusterComponentsRepoChangedEvent.getClusterId(), cluster, host);
+    }
   }
 
   @Subscribe
   public void onServiceComponentRecoveryChanged(ServiceComponentRecoveryChangedEvent event) throws AmbariException {
-    updateDataOfCluster(event.getClusterId());
+    long clusterId = event.getClusterId();
+    Cluster cluster = clusters.getCluster(clusterId);
+    for (ServiceComponentHost host : cluster.getServiceComponentHosts(event.getServiceName(), event.getComponentName())) {
+      updateDataOfHost(clusterId, cluster, host.getHost());
+    }
+  }
+
+  private void updateDataOfHost(long clusterId, Cluster cluster, Host host) throws AmbariException {
+    HostLevelParamsUpdateEvent hostLevelParamsUpdateEvent = new HostLevelParamsUpdateEvent(Long.toString(clusterId),
+            new HostLevelParamsCluster(
+                    m_ambariManagementController.get().retrieveHostRepositories(cluster, host),
+                    recoveryConfigHelper.getRecoveryConfig(cluster.getClusterName(), host.getHostName())));
+    hostLevelParamsUpdateEvent.setHostId(host.getHostId());
+    updateData(hostLevelParamsUpdateEvent);
   }
 
-  private void updateDataOfCluster(long clusterId) throws AmbariException {
+  @Subscribe
+  public void onServiceComponentRecoveryChanged(MaintenanceModeEvent event) throws AmbariException {
+    long clusterId = event.getClusterId();
     Cluster cluster = clusters.getCluster(clusterId);
-    for (Host host : cluster.getHosts()) {
-      HostLevelParamsUpdateEvent hostLevelParamsUpdateEvent = new HostLevelParamsUpdateEvent(Long.toString(clusterId),
-              new HostLevelParamsCluster(
-                      m_ambariManagementController.get().retrieveHostRepositories(cluster, host),
-                      recoveryConfigHelper.getRecoveryConfig(cluster.getClusterName(), host.getHostName())));
-      hostLevelParamsUpdateEvent.setHostId(host.getHostId());
-      updateData(hostLevelParamsUpdateEvent);
+    if (event.getHost() != null || event.getServiceComponentHost() != null) {
+      Host host = event.getHost() != null ? event.getHost() : cluster.getHost(event.getServiceComponentHost().getHostName());
+      updateDataOfHost(clusterId, cluster, host);
+    }
+    else if (event.getService() != null) {
+      for (String hostName : cluster.getService(event.getService().getName()).getServiceHosts()) {
```

Comment:
Is `cluster.getService(String)` call necessary instead of using `event.getService()` directly?